File: scala/scalafind/src/main/scala/scalafind/ArgTokenizer.scala
package scalafind

import org.json.{JSONArray, JSONException, JSONObject, JSONTokener}

import java.io.IOException
import java.nio.file.{Files, Path, Paths}
import scala.annotation.tailrec
import scala.jdk.CollectionConverters.*
import scala.util.matching.Regex

enum ArgTokenType:
  case Unknown, Bool, Str, Int, Long

case class ArgToken(name: String, tokenType: ArgTokenType, value: Any) {
}

class ArgTokenizer(options: List[FindOption]) {

  private var boolMap: Map[String, String] = Map.empty
  private var strMap: Map[String, String] = Map.empty
  private var intMap: Map[String, String] = Map.empty
  private var longMap: Map[String, String] = Map.empty

  options.foreach { opt =>
    opt.argType match {
      case ArgTokenType.Bool =>
        boolMap += (opt.longArg -> opt.longArg)
        if (opt.shortArg.isDefined) {
          boolMap += (opt.shortArg.get -> opt.longArg)
        }
      case ArgTokenType.Str =>
        strMap += (opt.longArg -> opt.longArg)
        if (opt.shortArg.isDefined) {
          strMap += (opt.shortArg.get -> opt.longArg)
        }
      case ArgTokenType.Int =>
        intMap += (opt.longArg -> opt.longArg)
        if (opt.shortArg.isDefined) {
          intMap += (opt.shortArg.get -> opt.longArg)
        }
      case ArgTokenType.Long =>
        longMap += (opt.longArg -> opt.longArg)
        if (opt.shortArg.isDefined) {
          longMap += (opt.shortArg.get -> opt.longArg)
        }
      case _ => // ignore
    }
  }

  private val longArgWithValRegex: Regex = "^--([a-zA-Z0-9-]+)=(.+)$".r
  private val longArgWithoutValRegex: Regex = "^--([a-zA-Z0-9-]+)$".r
  private val shortArgsRegex: Regex = "^-([a-zA-Z0-9]{2,})$".r
  private val shortArgRegex: Regex = "^-([a-zA-Z0-9])$".r

  private def getLongArg(argName: String): Option[String] = {
    if (boolMap.contains(argName)) {
      boolMap.get(argName)
    } else if (strMap.contains(argName)) {
      strMap.get(argName)
    } else if (intMap.contains(argName)) {
      intMap.get(argName)
    } else if (longMap.contains(argName)) {
      longMap.get(argName)
    } else {
      None
    }
  }

  private def argWantsVal(argName: String): Boolean = {
    strMap.contains(argName) || intMap.contains(argName) || longMap.contains(argName)
  }

  private def updateArgTokens(argName: String, argVal: Option[Any], argTokens: List[ArgToken]): List[ArgToken] = {
    val nextArgTokens =
      if (boolMap.contains(argName)) {
        argVal match {
          case None => List(ArgToken(argName, ArgTokenType.Bool, true))
          case Some(value) if value.isInstanceOf[Boolean] => List(ArgToken(argName, ArgTokenType.Bool, value))
          case Some(value) => throw new FindException("Invalid value for option: %s".format(argName))
        }
      } else if (strMap.contains(argName)) {
        argVal match {
          case None => throw new FindException("Missing value for option: %s".format(argName))
          case Some(value) if value.isInstanceOf[String] => List(ArgToken(argName, ArgTokenType.Str, value))
          case Some(value) if value.isInstanceOf[Iterable[?]] => value.asInstanceOf[Iterable[?]].map(v => ArgToken(argName, ArgTokenType.Str, v))
          case Some(value) if value.isInstanceOf[JSONArray] => value.asInstanceOf[JSONArray].toList.asScala.map(v => ArgToken(argName, ArgTokenType.Str, v))
          case Some(value) => throw new FindException("Invalid value for option: %s".format(argName))
        }
      } else if (intMap.contains(argName)) {
        argVal match {
          case None => throw new FindException("Missing value for option: %s".format(argName))
          case Some(value) if value.isInstanceOf[Int] => List(ArgToken(argName, ArgTokenType.Int, value))
          case Some(value) if value.isInstanceOf[Long] => List(ArgToken(argName, ArgTokenType.Int, value.asInstanceOf[Long].toInt))
          case Some(value) if value.isInstanceOf[String] => List(ArgToken(argName, ArgTokenType.Int, value.asInstanceOf[String].toInt))
          case Some(value) => throw new FindException("Invalid value for option: %s".format(argName))
        }
      } else if (longMap.contains(argName)) {
        argVal match {
          case None => throw new FindException("Missing value for option: %s".format(argName))
          case Some(value) if value.isInstanceOf[Long] => List(ArgToken(argName, ArgTokenType.Long, value))
          case Some(value) if value.isInstanceOf[Int] => List(ArgToken(argName, ArgTokenType.Long, value.asInstanceOf[Int].toLong))
          case Some(value) if value.isInstanceOf[String] => List(ArgToken(argName, ArgTokenType.Long, value.asInstanceOf[String].toLong))
          case Some(value) => throw new FindException("Invalid value for option: %s".format(argName))
        }
      } else {
        throw new FindException("Invalid option: %s".format(argName))
      }
    argTokens :++ nextArgTokens
  }

  @tailrec
  private def recArgTokensFromArgs(args: List[String], argTokens: List[ArgToken]): List[ArgToken] = args match {
    case Nil => argTokens
    case longArgWithValRegex(argName, argVal) :: tail =>
      recArgTokensFromArgs(tail, updateArgTokens(argName, Some(argVal), argTokens))
    case longArgWithoutValRegex(argName) :: tail =>
      val argVal = if (argWantsVal(argName) && tail.nonEmpty) {
        tail.headOption
      } else {
        None
      }
      val nextArgs = if (argVal.isDefined) tail.tail else tail
      recArgTokensFromArgs(nextArgs, updateArgTokens(argName, argVal, argTokens))
    case shortArgsRegex(shortArgs) :: tail =>
      val shortArgsList = shortArgs.toList.map(c => "-%s".format(c))
      recArgTokensFromArgs(shortArgsList ++ tail, argTokens)
    case shortArgRegex(shortArg) :: tail =>
      val longArg = getLongArg(shortArg) match {
        case Some(value) => value
        case None => throw new FindException("Invalid option: %s".format(shortArg))
      }
      val argVal = if (argWantsVal(shortArg) && tail.nonEmpty) {
        tail.headOption
      } else {
        None
      }
      val nextArgs = if (argVal.isDefined) tail.tail else tail
      recArgTokensFromArgs(nextArgs, updateArgTokens(longArg, argVal, argTokens))
    case arg :: tail =>
      recArgTokensFromArgs(tail, updateArgTokens("path", Some(arg), argTokens))
  }

  def tokenizeArgs(args: Array[String]): List[ArgToken] = {
    recArgTokensFromArgs(args.toList, List.empty[ArgToken])
  }

  def tokenizeJson(json: String): List[ArgToken] = {
    @tailrec
    def recArgTokensFromJson(keys: List[String], jsonObject: JSONObject, argTokens: List[ArgToken]): List[ArgToken] = keys match {
      case Nil => argTokens
      case k :: ks =>
        val v = jsonObject.get(k)
        recArgTokensFromJson(ks, jsonObject, updateArgTokens(k, Option(v), argTokens))
    }

    val jsonObject = new JSONObject(new JSONTokener(json))
    // keys are sorted so that output is consistent across all versions
    val keys = jsonObject.keySet().asScala.toList.sorted
    recArgTokensFromJson(keys, jsonObject, List.empty[ArgToken])
  }

  def tokenizeFile(filePath: String): List[ArgToken] = {
    val path: Path = FileUtil.expandPath(Paths.get(filePath))
    if (!Files.exists(path)) {
      throw new FindException("Settings file not found: %s".format(filePath))
    }
    if (!filePath.endsWith(".json")) {
      throw new FindException("Invalid settings file (must be JSON): %s".format(filePath))
    }
    try {
      val json: String = FileUtil.getPathContents(path)
      tokenizeJson(json)
    } catch {
      case e: IOException =>
        throw new FindException("Error reading settings file: %s".format(filePath))
      case e: JSONException =>
        throw new FindException("Error parsing JSON file: %s".format(filePath))
    }
  }
}

File: kotlin/ktfind/lib/src/main/kotlin/ktfind/ArgTokenizer.kt
package ktfind

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.json.JSONTokener
import java.io.FileNotFoundException
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

enum class ArgTokenType {
    UNKNOWN, BOOL, STR, INT, LONG
}

data class ArgToken(val name: String, val type: ArgTokenType, val value: Any) {
}

class ArgTokenizer(options: List<Option>) {

    private var boolMap = mutableMapOf<String, String>()
    private var strMap = mutableMapOf<String, String>()
    private var intMap = mutableMapOf<String, String>()
    private var longMap = mutableMapOf<String, String>()

    private val longArgWithValRegex = Regex("^--([a-zA-Z0-9-]+)=(.*)$")
    private val longArgWithoutValRegex = Regex("^--([a-zA-Z0-9-]+)$")
    private val shortArgsRegex = Regex("^-([a-zA-Z0-9]{2,})$")
    private val shortArgRegex = Regex("^-([a-zA-Z0-9])$")

    init {
        options.forEach { option ->
            when (option.argType) {
                ArgTokenType.BOOL -> {
                    boolMap[option.longArg] = option.longArg
                    option.shortArg?.let { boolMap[it] = option.longArg }
                }

                ArgTokenType.STR -> {
                    strMap[option.longArg] = option.longArg
                    option.shortArg?.let { strMap[it] = option.longArg }
                }

                ArgTokenType.INT -> {
                    intMap[option.longArg] = option.longArg
                    option.shortArg?.let { intMap[it] = option.longArg }
                }

                ArgTokenType.LONG -> {
                    longMap[option.longArg] = option.longArg
                    option.shortArg?.let {
                        longMap[it] = option.longArg
                    }
                }

                else -> {
                    throw FindException("Invalid option: ${option.longArg}")
                }
            }
        }
    }

    private fun updateArgTokens(argName: String, argVal: Any?, argTokens: List<ArgToken>): List<ArgToken> {
        val nextArgTokens: List<ArgToken> =
            if (boolMap.containsKey(argName)) {
                when (argVal) {
                    null -> {
                        listOf(ArgToken(boolMap[argName]!!, ArgTokenType.BOOL, true))
                    }
                    is Boolean -> {
                        listOf(ArgToken(boolMap[argName]!!, ArgTokenType.BOOL, argVal))
                    }
                    else -> {
                        throw FindException("Invalid value for option: $argName")
                    }
                }
            } else if (strMap.containsKey(argName)) {
                when (argVal) {
                    null -> {
                        throw FindException("Missing value for option: $argName")
                    }
                    is String -> {
                        listOf(ArgToken(strMap[argName]!!, ArgTokenType.STR, argVal))
                    }
                    is Collection<*> -> {
                        if (argVal.any({ v -> v !is String })) {
                            throw FindException("Invalid value for option: $argName")
                        }
                        argVal.map { v -> ArgToken(strMap[argName]!!, ArgTokenType.STR, v as String) }.toList()
                    }
                    is JSONArray -> {
                        if (argVal.any({ v -> v !is String })) {
                            throw FindException("Invalid value for option: $argName")
                        }
                        argVal.map { v -> ArgToken(strMap[argName]!!, ArgTokenType.STR, v as String) }.toList()
                    }
                    else -> {
                        throw FindException("Invalid value for option: $argName")
                    }
                }
            } else if (intMap.containsKey(argName)) {
                when (argVal) {
                    null -> {
                        throw FindException("Missing value for option: $argName")
                    }
                    is Int -> {
                        listOf(ArgToken(intMap[argName]!!, ArgTokenType.INT, argVal))
                    }
                    is Long -> {
                        listOf(ArgToken(intMap[argName]!!, ArgTokenType.INT, argVal.toInt()))
                    }
                    is String -> {
                        listOf(ArgToken(intMap[argName]!!, ArgTokenType.INT, argVal.toInt()))
                    }
                    else -> {
                        throw FindException("Invalid value for option: $argName")
                    }
                }
            } else if (longMap.containsKey(argName)) {
                when (argVal) {
                    null -> {
                        throw FindException("Missing value for option: $argName")
                    }
                    is Long -> {
                        listOf(ArgToken(longMap[argName]!!, ArgTokenType.LONG, argVal))
                    }
                    is Int -> {
                        listOf(ArgToken(longMap[argName]!!, ArgTokenType.LONG, argVal.toLong()))
                    }
                    is String -> {
                        listOf(ArgToken(intMap[argName]!!, ArgTokenType.INT, argVal.toLong()))
                    }
                    else -> {
                        throw FindException("Invalid value for option: $argName")
                    }
                }
            } else {
                throw FindException("Invalid option: $argName")
            }
        return argTokens + nextArgTokens
    }

    fun tokenizeArgs(args: Array<String>): List<ArgToken> {
        fun recArgTokensFromArgs(args: List<String>, argTokens: List<ArgToken>): List<ArgToken> {
            if (args.isEmpty()) return argTokens
            val nextArg = args.first()
            if (nextArg matches longArgWithValRegex) {
                val match = longArgWithValRegex.matchEntire(nextArg)
                val longArg = match!!.groupValues[1]
                val argVal = match.groupValues[2]
                return recArgTokensFromArgs(args.drop(1), updateArgTokens(longArg, argVal, argTokens))
            }
            if (nextArg matches longArgWithoutValRegex) {
                val match = longArgWithoutValRegex.matchEntire(nextArg)
                val longArg = match!!.groupValues[1]
                val argVal: String? =
                    if ((strMap.containsKey(longArg)
                                || intMap.containsKey(longArg)
                                || longMap.containsKey(longArg))
                        && args.size > 1) {
                        args[1]
                    } else {
                        null
                    }
                val nextArgs =
                    if (argVal != null) {
                        args.drop(2)
                    } else {
                        args.drop(1)
                    }
                return recArgTokensFromArgs(nextArgs, updateArgTokens(longArg, argVal, argTokens))
            }
            if (nextArg matches shortArgsRegex) {
                val match = shortArgsRegex.matchEntire(nextArg)
                val shortArgs = match!!.groupValues[1]
                val shortArgList = shortArgs.map { c -> "-$c" }
                val nextArgs = shortArgList + args.drop(1)
                return recArgTokensFromArgs(nextArgs, argTokens)
            }
            if (nextArg matches shortArgRegex) {
                val match = shortArgRegex.matchEntire(nextArg)
                val shortArg = match!!.groupValues[1]
                val longArg: String? =
                    if (boolMap.containsKey(shortArg)) boolMap[shortArg]
                    else if (strMap.containsKey(shortArg)) strMap[shortArg]
                    else if (intMap.containsKey(shortArg)) intMap[shortArg]
                    else if (longMap.containsKey(shortArg)) longMap[shortArg]
                    else null
                if (longArg == null) {
                    throw FindException("Invalid option: $shortArg")
                }
                val argVal: String? =
                    if ((strMap.containsKey(longArg)
                                || intMap.containsKey(longArg)
                                || longMap.containsKey(longArg))
                        && args.size > 1) {
                        args[1]
                    } else {
                        null
                    }
                val nextArgs =
                    if (argVal != null) {
                        args.drop(2)
                    } else {
                        args.drop(1)
                    }
                return recArgTokensFromArgs(nextArgs, updateArgTokens(longArg, argVal, argTokens))
            }
            return recArgTokensFromArgs(args.drop(1), argTokens + ArgToken("path", ArgTokenType.STR, nextArg))
        }
        return recArgTokensFromArgs(args.toList(), listOf())
    }

    fun tokenizeJson(json: String): List<ArgToken> {
        fun recTokenizeJson(keys: List<String>, jsonObject: JSONObject, argTokens: List<ArgToken>): List<ArgToken> {
            return if (keys.isEmpty()) argTokens
            else {
                val k = keys.first()
                val v = jsonObject.get(k)
                recTokenizeJson(keys.drop(1), jsonObject, updateArgTokens(k, v, argTokens))
            }
        }
        try {
            val jsonObject = JSONObject(JSONTokener(json))
            // keys are sorted so that output is consistent across all versions
            val keys = jsonObject.keySet().toList().sorted()
            return recTokenizeJson(keys, jsonObject, listOf())
        } catch (_: JSONException) {
            throw FindException("Unable to parse JSON settings")
        }
    }

    fun tokenizeFile(filePath: String): List<ArgToken> {
        val path = FileUtil.expandPath(Paths.get(filePath));
        if (!Files.exists(path)) {
            throw FindException("Settings file not found: $filePath")
        }
        if (!filePath.endsWith(".json")) {
            throw FindException("Invalid settings file (must be JSON): $filePath")
        }
        try {
            val json = path.toFile().readText()
            return tokenizeJson(json)
        } catch (_: FileNotFoundException) {
            throw FindException("Settings file not found: $filePath")
        } catch (_: IOException) {
            throw FindException("IOException reading settings file: $filePath")
        } catch (_: JSONException) {
            throw FindException("Unable to parse JSON in settings file: $filePath")
        }
    }
}
